src/ui/marker.js

```diff
@@ -444,35 +448,68 @@ export default class Marker extends Evented {
         return this;
     }
 
-    _updateOcclusion() {
-        if (!this._occlusionTimer) {
-            this._occlusionTimer = setTimeout(this._onOcclusionTimer.bind(this), 60);
-        }
-    }
+    _evaluateOpacity(requestAnimationFrame: boolean = false) {
+        const position = this._pos ? this._pos.sub(this._transformedOffset()) : null;
 
-    _clearOcclusionTimer() {
-        if (this._occlusionTimer) {
-            clearTimeout(this._occlusionTimer);
-            this._occlusionTimer = null;
+        if (!this._withinScreenBounds(position)) {
+            this._fadeTimer = null;
+            return;
         }
-    }
 
-    _onOcclusionTimer() {
-        const tr = this._map.transform;
-        const pos = this._pos ? this._pos.sub(this._transformedOffset()) : null;
-        if (pos && pos.x >= 0 && pos.x < tr.width && pos.y >= 0 && pos.y < tr.height) {
-            // calculate if occluded.
-            const raycastLoc = this._map.unproject(pos);
+        const mapLocation = this._map.unproject(position);
+
+        let terrainOccluded = false;
+        if (this._map.getTerrain()) {
             const camera = this._map.getFreeCameraOptions();
             if (camera.position) {
                 const cameraPos = camera.position.toLngLat();
-                const raycastDistance = cameraPos.distanceTo(raycastLoc);
-                const posDistance = cameraPos.distanceTo(this._lngLat);
-                const occluded = raycastDistance < posDistance * 0.9;
-                this._element.classList.toggle('mapboxgl-marker-occluded', occluded);
+                // the distance to the marker lat/lng + marker offset location
+                const offsetDistance = cameraPos.distanceTo(mapLocation);
+                const distance = cameraPos.distanceTo(this._lngLat);
+                terrainOccluded = offsetDistance < distance * 0.9;
+            }
+        }
+
+        const fogOpacity = this._map.queryFogOpacity(mapLocation);
+        const opacity = (1.0 - fogOpacity) * (terrainOccluded ? TERRAIN_OCCLUDED_OPACITY : 1.0);
+
+        const occluded = opacity === 0;
+        const occludedLow = opacity > 0 && opacity <= 0.25;
+        const occludedMid = opacity > 0.25 && opacity <= 0.5;
+        const occludedHigh = opacity > 0.5 && opacity <= 0.75;
+
+        const updateStyle = () => {
+            this._element.classList.toggle('mapboxgl-marker-occluded', occluded);
```

Comment:
We were initially modifying the html style directly ([reference commit](https://github.com/mapbox/mapbox-gl-js/pull/10516/commits/0683f69143c13403465067af696cef1904523479#diff-12738a6ec6570cfc78a8c4e09bec0ff1f9f8762b3372ff9dd80c4fdc9436f24cR480)) but from a request from @arindam1993 https://github.com/mapbox/mapbox-gl-js/pull/10516#pullrequestreview-624513275, we switched to having a few stops. Mutating style opacity was considered a potential bottleneck for performance reasons. I have no strong opinion as I'm not very well informed on that topic.

Reply:
It was for:
1. Minimizing dom updates, and letting CSS transitions fill in for smoothing between steps
2. Since users can set custom content inside the marker, adding a class at the root DOM element makes everything downstream inherit it clearly.
3. Users can also override the CSS allowing for customizability of the opacity steps without needing more API methods on the markers.

Reply:
1. I think changing the opacity once per frame should be ok. We already change the position once per frame.
2. I might be misunderstanding, but doesn't setting opacity directly affect child elements the same as setting the class?
3. the new classes are the new public api

Reply:
Making the classes the public API instead of having to set something in JS was the intention.

And we cant update opacity every frame, the raycasting through terrain isn't fast enough to run every frame for a large number of markers, so it runs at a reduced rate hence the need for CSS transitions to fill in the gaps smoothly